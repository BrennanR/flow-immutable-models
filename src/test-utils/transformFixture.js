// @flow
import fs from 'fs';
import jscodeshift from 'jscodeshift';
import transform from '../transform';

export default function transformFile(filename: string): Promise<string> {
  return new Promise((resolve, reject) => {
    const file = fs.readFile(filename, (err, source) => {
      if (err) {
        reject(err);
        return;
      }
      const output = transform({
        path: file,
        source: source.toString(),
      }, {
        jscodeshift,
        stats: () => {},
      }, {});
      resolve(output.toString());
    });
  });
}
