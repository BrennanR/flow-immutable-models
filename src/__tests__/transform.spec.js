import path from 'path';
import transformFixture from '../test-utils/transformFixture';

async function snap(filename: string): string {
  const output = await transformFixture(path.join(__dirname, 'fixtures', filename));
  expect(output).toMatchSnapshot();
}

describe('transform', () => {
  it('adds default class for exported typed object', async () => {
    await snap('exportedType.js');
  });

  it('initilizes Immutable lists and maps', async () => {
    await snap('immutableProperties.js');
  });

  it('initilizes referenced types', async () => {
    await snap('referenceProperties.js');
  });

  it('initilizes arrays without new operator', async () => {
    await snap('arrayType.js');
  });

  it('does not instantiate inline non-class props', async () => {
    await snap('nonClassInlineTypes.js');
  });

  it('does not instantiate imported types', async () => {
    await snap('nonClassImportedTypes.js');
  });
});
