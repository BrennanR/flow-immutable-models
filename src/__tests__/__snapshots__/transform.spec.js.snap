exports[`transform adds default class for exported typed object 1`] = `
"// @flow
import type { BarInterface } from \'./bar\';
import { Bar } from \'./bar\';
import ImmutableModel from \'../../../src/ImmutableModel\';

export type FooInterface = {
  maybeStr: ?string,
  str: string,
  maybeNum: ?number,
  num: number,
  maybeObj: ?Object,
  obj: Object,
  maybeBar: ?BarInterface,
  bar: BarInterface,
  maybeFn: ?Function,
  fn: Function,
};

export class Foo extends ImmutableModel {
  static fromJS(state: FooInterface): Foo {
    return this.fromImmutable(Immutable.fromJS(state));
  }

  static fromImmutable(immutableState: Immutable.Map<string, any>): Foo {
    return new Foo(immutableState);
  }

  initialize(): Immutable.Map<string, any> {
    let state = Immutable.Map();
    state = state.set(\'bar\', new Bar());
    return state;
  }

  get maybeStr(): ?string {
    return this._state.get(\'maybeStr\');
  }

  setMaybeStr(maybeStr: ?string): Foo {
    return new Foo(this._state.set(\'maybeStr\'));
  }

  get str(): string {
    return this._state.get(\'str\');
  }

  setStr(str: string): Foo {
    return new Foo(this._state.set(\'str\'));
  }

  get maybeNum(): ?number {
    return this._state.get(\'maybeNum\');
  }

  setMaybeNum(maybeNum: ?number): Foo {
    return new Foo(this._state.set(\'maybeNum\'));
  }

  get num(): number {
    return this._state.get(\'num\');
  }

  setNum(num: number): Foo {
    return new Foo(this._state.set(\'num\'));
  }

  get maybeObj(): ?Object {
    return this._state.get(\'maybeObj\');
  }

  setMaybeObj(maybeObj: ?Object): Foo {
    return new Foo(this._state.set(\'maybeObj\'));
  }

  get obj(): Object {
    return this._state.get(\'obj\');
  }

  setObj(obj: Object): Foo {
    return new Foo(this._state.set(\'obj\'));
  }

  get maybeBar(): ?Bar {
    return this._state.get(\'maybeBar\');
  }

  setMaybeBar(maybeBar: ?Bar): Foo {
    return new Foo(this._state.set(\'maybeBar\'));
  }

  get bar(): Bar {
    return this._state.get(\'bar\');
  }

  setBar(bar: Bar): Foo {
    return new Foo(this._state.set(\'bar\'));
  }

  get maybeFn(): ?Function {
    return this._state.get(\'maybeFn\');
  }

  setMaybeFn(maybeFn: ?Function): Foo {
    return new Foo(this._state.set(\'maybeFn\'));
  }

  get fn(): Function {
    return this._state.get(\'fn\');
  }

  setFn(fn: Function): Foo {
    return new Foo(this._state.set(\'fn\'));
  }
}
"
`;

exports[`transform initilizes Immutable lists and maps 1`] = `
"// @flow
import * as Immutable from \'immutable\';
import type { BarInterface } from \'./bar\';
import { Bar } from \'./bar\';
import ImmutableModel from \'../../../src/ImmutableModel\';

export type FooInterface = {
  list: Immutable.List<any>,
  maybeList: ?Immutable.List<BarInterface>,
  map: Immutable.Map<any, any>,
  maybeMap: ?Immutable.Map<string, BarInterface>,
};

export class Foo extends ImmutableModel {
  static fromJS(state: FooInterface): Foo {
    return this.fromImmutable(Immutable.fromJS(state));
  }

  static fromImmutable(immutableState: Immutable.Map<string, any>): Foo {
    return new Foo(immutableState);
  }

  initialize(): Immutable.Map<string, any> {
    let state = Immutable.Map();
    state = state.set(\'list\', Immutable.List());
    state = state.set(\'map\', Immutable.Map());
    return state;
  }

  get list(): Immutable.List<any> {
    return this._state.get(\'list\');
  }

  setList(list: Immutable.List<any>): Foo {
    return new Foo(this._state.set(\'list\'));
  }

  get maybeList(): ?Immutable.List<Bar> {
    return this._state.get(\'maybeList\');
  }

  setMaybeList(maybeList: ?Immutable.List<Bar>): Foo {
    return new Foo(this._state.set(\'maybeList\'));
  }

  get map(): Immutable.Map<any, any> {
    return this._state.get(\'map\');
  }

  setMap(map: Immutable.Map<any, any>): Foo {
    return new Foo(this._state.set(\'map\'));
  }

  get maybeMap(): ?Immutable.Map<string, Bar> {
    return this._state.get(\'maybeMap\');
  }

  setMaybeMap(maybeMap: ?Immutable.Map<string, Bar>): Foo {
    return new Foo(this._state.set(\'maybeMap\'));
  }
}
"
`;

exports[`transform initilizes referenced types 1`] = `
"// @flow
import type { BarInterface } from \'./bar\';
import { Bar } from \'./bar\';
import ImmutableModel from \'../../../src/ImmutableModel\';

export type FooInterface = {
  ref: BarInterface,
  maybeRef: ?BarInterface,
  str: string,
};

export class Foo extends ImmutableModel {
  static fromJS(state: FooInterface): Foo {
    return this.fromImmutable(Immutable.fromJS(state));
  }

  static fromImmutable(immutableState: Immutable.Map<string, any>): Foo {
    return new Foo(immutableState);
  }

  initialize(): Immutable.Map<string, any> {
    let state = Immutable.Map();
    state = state.set(\'ref\', new Bar());
    return state;
  }

  get ref(): Bar {
    return this._state.get(\'ref\');
  }

  setRef(ref: Bar): Foo {
    return new Foo(this._state.set(\'ref\'));
  }

  get maybeRef(): ?Bar {
    return this._state.get(\'maybeRef\');
  }

  setMaybeRef(maybeRef: ?Bar): Foo {
    return new Foo(this._state.set(\'maybeRef\'));
  }

  get str(): string {
    return this._state.get(\'str\');
  }

  setStr(str: string): Foo {
    return new Foo(this._state.set(\'str\'));
  }
}
"
`;
